### PR TITLE
Improve models memory usage

### LINKED ISSUE #348: Determine if using slots for models improves memory usage.
We should look into if we can use `__slots__` in our models to decrease memory usage in mopidy. Change should be conditional on testing/benchmarking that shows that the change is worth it.

### PR DESCRIPTION
**Pull request by adamcik**
Fixes #348

### CODE REVIEW

File: mopidy/models.py [hunk 1]
```diff
@@ -123,18 +133,32 @@ def validate(self, value):
...
             if isinstance(value, Field):
-                attrs['_fields'].append(key)
+                fields[key] = '_' + key
                 value._name = key
-        attrs['_fields'].sort()
-        return super(FieldOwner, cls).__new__(cls, name, bases, attrs)
+
+        attrs['_fields'] = fields
+        attrs['_instances'] = weakref.WeakValueDictionary()
+        attrs['__slots__'] = ['_' + field for field in fields]
```
Comment (jodal) [inline mopidy/models.py]: What about `= fields.values()`?

File: mopidy/models.py [hunk 2]
```diff
@@ -264,13 +313,13 @@ def model_json_decoder(dct):
...
+        models = {c.__name__: c for c in ImmutableObject.__subclasses__()}
         model_name = dct.pop('__model__')
-        cls = globals().get(model_name, None)
-        if issubclass(cls, ImmutableObject):
+        if model_name in models:
             kwargs = {}
             for key, value in dct.items():
                 kwargs[key] = value
-            return cls(**kwargs)
+            return models[model_name](**kwargs)
```
Comment (jodal) [inline mopidy/models.py]: I know the previous three lines have not been changed by this PR, but can't they simply be removed and `**kwargs` be changed to `**dct`?

**Comment by jodal** [comment]
That was all :-)

File: mopidy/models.py (hunk 1 above)
Comment (adamcik) [inline mopidy/models.py]: Yeah, right, that was the original plan that seems to have been lost.

**Comment by adamcik** [comment]
Done.

**Pull request merged**